rasa/nlu/featurizers/featurizer.py

```diff
@@ -1,48 +1,142 @@
-import numpy as np
-from typing import Text, Optional, Dict, Any
+from __future__ import annotations
+from abc import abstractmethod
+from collections import Counter
+from rasa.nlu.tokenizers.tokenizer import Tokenizer
+from typing import Generic, Iterable, List, Text, Optional, Dict, Any, TypeVar
 
 from rasa.nlu.constants import FEATURIZER_CLASS_ALIAS
-from rasa.nlu.components import Component
-from rasa.utils.tensorflow.constants import MEAN_POOLING, MAX_POOLING
+from rasa.shared.nlu.training_data.features import Features
+from rasa.shared.nlu.training_data.message import Message
+from rasa.shared.nlu.training_data.training_data import TrainingData
+from rasa.shared.exceptions import InvalidConfigException
+from rasa.shared.nlu.constants import FEATURE_TYPE_SENTENCE, FEATURE_TYPE_SEQUENCE
 
+# TODO: remove after all featurizers have been migrated
+from rasa.nlu.featurizers._featurizer import (
+    Featurizer,
+    SparseFeaturizer,
+    DenseFeaturizer,
+)
 
-class Featurizer(Component):
-    def __init__(self, component_config: Optional[Dict[Text, Any]] = None) -> None:
-        if not component_config:
-            component_config = {}
+Featurizer = Featurizer
+SparseFeaturizer = SparseFeaturizer
+DenseFeaturizer = DenseFeaturizer
 
-        # makes sure the alias name is set
-        # Necessary for `unique_name` to be defined
-        self.component_config = component_config
-        component_config.setdefault(FEATURIZER_CLASS_ALIAS, self.unique_name)
+FeatureType = TypeVar("FeatureType")
 
-        super().__init__(component_config)
 
+class Featurizer2(Generic[FeatureType]):
+    """Base class for all featurizers."""
 
-class DenseFeaturizer(Featurizer):
     @staticmethod
-    def _calculate_sentence_features(
-        features: np.ndarray, pooling_operation: Text
-    ) -> np.ndarray:
-        # take only non zeros feature vectors into account
-        non_zero_features = np.array([f for f in features if f.any()])
+    def get_default_config() -> Dict[Text, Any]:
+        """Returns the component's default config."""
+        return {FEATURIZER_CLASS_ALIAS: None}
 
-        # if features are all zero just return a vector with all zeros
-        if non_zero_features.size == 0:
-            return np.zeros([1, features.shape[-1]])
+    def __init__(self, config: Optional[Dict[Text, Any]] = None) -> None:
+        """Instantiates a new featurizer.
 
-        if pooling_operation == MEAN_POOLING:
-            return np.mean(non_zero_features, axis=0, keepdims=True)
+        Args:
+          config: configuration
+        """
+        super().__init__()
+        self.validate_config(config)
+        self._config = config
+        # TODO: we can't use COMPONENT_INDEX anymore to create a unique name
+        # on the fly --> take care of that in graph validation
+        self._identifier = self._config.get(FEATURIZER_CLASS_ALIAS, None)
```

Comment:
We can get the `node_name` from the `ExecutionContext`

Reply:
Ah, nice - will use that as a default to overwrite the configured alias